### PR TITLE
Move TraceSchedulingAutoConfiguration @Conditional for optional AspectJ

### PR DESCRIPTION
**Pull request by timtebeek**
Fixes spring-cloud/spring-cloud-sleuth#1621

With the conditional in the original location the tests fail;
With the conditional in the new location the tests pass.

### CODE REVIEW

**Comment by timtebeek** [comment]
@snicoll you mentioned
> I should also note that this annotation is overriding what the AOP auto-configuration in Spring Boot does. Nothing wrong about doing that but it's doing that unilaterally in an auto-configuration so I don't think that was the original intent.

Would that be a reason to remove `@EnableAspectJAutoProxy` from [TraceSchedulingAutoConfiguration](https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfiguration.java#L43) altogether?

**Comment by timtebeek** [comment]
Happy to help! Appreciate the work that's gone into Sleuth already, so glad to be able to contribute a bit back to the project.

It seems Stéphane is referencing the Spring Boot [AopAutoConfiguration](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/aop/AopAutoConfiguration.java), so let's wait to see if he would like to see any further changes to `TraceSchedulingAutoConfiguration` before we finish up.

**Comment by snicoll** [comment]
This PR is one of the option discussed in the original issue but we need to figure out what the intention originally was and I don't have that information. 

> Would that be a reason to remove @EnableAspectJAutoProxy from TraceSchedulingAutoConfiguration altogether?

I think so.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfiguration.java [hunk 1]
```diff
@@ -41,14 +41,14 @@
  */
 @Configuration(proxyBeanMethods = false)
 @EnableAspectJAutoProxy
```
Comment (timtebeek) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfiguration.java:43]:
```suggestion
```

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.scheduling;
+
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.junit.jupiter.api.Test;
+
+import org.springframework.aop.aspectj.annotation.AnnotationAwareAspectJAutoProxyCreator;
```
Comment (timtebeek) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfigurationTest.java:22]:
```suggestion
```

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+			.withConfiguration(AutoConfigurations.of(
+					TraceAutoConfiguration.class,
+					TraceSchedulingAutoConfiguration.class));
+
+	@Test
+	void shoud_create_TraceSchedulingAspect() {
+		this.contextRunner
+				.run(context -> assertThat(context)
+						.hasSingleBean(TraceSchedulingAspect.class)
+						.hasSingleBean(AnnotationAwareAspectJAutoProxyCreator.class));
```
Comment (timtebeek) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfigurationTest.java:42]:
```suggestion
						.hasSingleBean(TraceSchedulingAspect.class));
```

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	@Test
+	void shoud_not_create_TraceSchedulingAspect_without_aspectJ() {
+		this.contextRunner
+				.withClassLoader(new FilteredClassLoader(ProceedingJoinPoint.class))
+				.run(context -> assertThat(context)
+						.doesNotHaveBean(TraceSchedulingAspect.class)
+						.doesNotHaveBean(AnnotationAwareAspectJAutoProxyCreator.class));
```
Comment (timtebeek) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAutoConfigurationTest.java:51]:
```suggestion
						.hasSingleBean(TraceSchedulingAspect.class));
```

**Comment by timtebeek** [comment]
> This PR is one of the option discussed in the original issue but we need to figure out what the intention originally was and I don't have that information. Edit: ah I just saw Adrian approved the PR so forget it :)
> 
> > Would that be a reason to remove @EnableAspectJAutoProxy from TraceSchedulingAutoConfiguration altogether?
> 
> I think so. if I remember correctly, that's not the only place where this is invoked so if we decide to remove it, there might be other places where a similar change should be applied.

Could not find any other mentions of `EnableAspectJAutoProxy` in spring-cloud-sleuth, so if you want it removed @snicoll you can commit the suggestions above. From your message is wasn't clear cut if that's what you intended, so I've added it as review suggestions for a quick fix within Github. Hope it helps!

**Comment by adriancole** [comment]
I'll merge this locally as it is only checkstyle. thanks!

**Comment by adriancole** [comment]
into master as bf9cfe5cc I'll port this and the other changes that went to master recently back to 2.2.x Thanks @timtebeek and @snicoll

**Pull request closed without merging**